src/queue.jl

```diff
@@ -6,31 +6,57 @@ Each queue is uniquely associated with an agent.
 """
 mutable struct HSAQueue
     agent::HSAAgent
-    queue::Ref{Ptr{HSA.Queue}}
+    queue::Ptr{HSA.Queue}
     active::Bool
 end
-
+const QUEUES = WeakKeyDict{Ptr{HSA.Queue},HSAQueue}()
 const DEFAULT_QUEUES = IdDict{HSAAgent,HSAQueue}()
 
+function queue_error_handler(status::HSA.Status, _queue::Ptr{HSA.Queue}, data::Ptr{Cvoid})::Nothing
+    # TODO: From which thread are we calling this?
+    if status !== HSA.STATUS_SUCCESS then
+        queue = QUEUES[_queue]
+        queue.active = false
+        @assert DEFAULT_QUEUES[queue.agent] == queue # @jpsamaroo who gurantuees this 1:1 relation?
+        delete!(DEFAULT_QUEUES, queue.agent)
+        lock(RT_LOCK) do
+            delete!(_active_kernels, queue)
+        end
+        # do we need to call `HSA.queue_destroy`?
+        error("Queue at $(queue) inactivated due to async error:\n\t$(description(HSAError(status)))\n")
+    end
+
+    return nothing
+end
+
 function HSAQueue(agent::HSAAgent)
     queue_size = Ref{UInt32}(0)
     getinfo(agent.agent, HSA.AGENT_INFO_QUEUE_MAX_SIZE, queue_size) |> check
     @assert queue_size[] > 0
-    queue = HSAQueue(agent, Ref{Ptr{HSA.Queue}}(), true)
+
+    r_queue = Ref{Ptr{HSA.Queue}}()
     HSA.queue_create(agent.agent, queue_size[], HSA.QUEUE_TYPE_MULTI,
-                     C_NULL, C_NULL, typemax(UInt32), typemax(UInt32),
-                     queue.queue) |> check
+                     @cfunction(queue_error_handler, Cvoid, (HSA.Status, Ptr{HSA.Queue}, Ptr{Cvoid})), C_NULL,
+                     typemax(UInt32), typemax(UInt32), r_queue) |> check    
+
+    queue = HSAQueue(agent, r_queue[], true)
+    @assert !haskey(QUEUES, queue.queue)
+    QUEUES[queue.queue] = queue
 
     lock(RT_LOCK) do
         _active_kernels[queue] = Vector{AMDGPU.RuntimeEvent{AMDGPU.HSAStatusSignal}}()
     end
 
     hsaref!()
     finalizer(queue) do queue
+        if haskey(QUEUES, queue.queue) then
+            delete!(QUEUES, queue.queue)
+        end
         if queue.active
-            HSA.queue_destroy(queue.queue[]) |> check
+            HSA.queue_destroy(queue.queue) |> check
         end
         lock(RT_LOCK) do
+            # XXX: What if someone already did this?
```

Comment:
We should probably ensure that if `queue.active == false`, it's already been deleted.